### PR TITLE
Add test for zero stray output from java async-proviler

### PR DESCRIPTION
**Pull request by marcin-ol**
## Description
Make sure that async-profiler generates no output to profiled process stdout/ stderr streams.

## Related Issue
#649 

## Motivation and Context
We don't need any output from async-profiler to appear in profiled process output streams. We don't need it and want to avoid it.  
Recently few commits were merged which removed some print statements reaching process outputs.

## How Has This Been Tested?
Testcase was added to `test_java` testsuite.

### CODE REVIEW

File: tests/test_java.py [hunk 1]
```diff
@@ -718,3 +718,42 @@ def test_java_frames_include_no_semicolons(
...
+    with make_java_profiler(
+        storage_dir=str(tmp_path),
+        duration=3,
+        frequency=999,
+    ) as profiler:
+        snapshot_pid_profile(profiler, application_pid)
+        textout = application_docker_container.logs(stdout=True, stderr=False).decode()
+        texterr = application_docker_container.logs(stdout=False, stderr=True).decode()
+        # output from Fibonacci should be the only output in stdout
+        assert textout.splitlines() == ["Fibonacci thread starting"]
```
Comment (Jongy) [inline tests/test_java.py:757]:
```suggestion
        assert textout == "Fibonacci thread starting\n"
```

simpler? Same for stderr

File: tests/test_java.py [hunk 2]
```diff
@@ -718,3 +718,42 @@ def test_java_frames_include_no_semicolons(
...
+    application_pid: int,
+    application_docker_container: Container,
+    monkeypatch: MonkeyPatch,
+) -> None:
+    # save original stop function
+    stop_async_profiler = AsyncProfiledProcess.stop_async_profiler
+
+    # replace async profiler stop routine to trigger flushing standard streams
+    def flush_output_and_stop_async_profiler(self: AsyncProfiledProcess, *args, **kwargs) -> str:
+        # Call 'version' and 'check' actions on async-profiler to make sure writes to stdout, stderr are flushed
```
Comment (Jongy) [inline tests/test_java.py:736]: Why does it cause flushing?

Anyway - a simpler solution: run the container, do a single profile, then stop the container; I suppose Java handles SIGTERM gracefully, so you will get the logs properly.

**Review by Jongy — CHANGES_REQUESTED** [review]
Done reviewing.

Please run a negative test - i.e a test where you change the AP reference to your fork, which removes e.g https://github.com/Granulate/async-profiler/commit/35531a45a197048a450aca38386d144dfa1060bb, so it proves that the test now fails.

File: tests/test_java.py (hunk 2 above)
Comment (marcin-ol) [reply tests/test_java.py:736]: That's true - java applications flush stdout by default (they trigger autoflush by default).  
async-profiler on the other hand is c++, and doesn't flush by default after each write to output stream.

I have observed that `version` action does trigger flush on standard output: ([profiler.cpp#L1548](https://github.com/granulate/async-profiler/blob/master/src/profiler.cpp#L1548)):
```cpp
        case ACTION_VERSION:
            out << PROFILER_VERSION;
            out.flush();
            break;
```
(this code is still there in the upstream, btw).

File: tests/test_java.py (hunk 1 above)
Comment (marcin-ol) [reply tests/test_java.py:757]: Yes.

File: tests/test_java.py (hunk 2 above)
Comment (Jongy) [reply tests/test_java.py]: I see. Good point. Please keep the code doing the `version` call and elaborate in a comment on why it's needed (you can also give a github permalink to this code which it the clear explanation why we want to run `version`).

**Review by Jongy — CHANGES_REQUESTED** [review]
Done reviewing

File: tests/test_java.py [hunk 3]
```diff
@@ -718,3 +718,46 @@ def test_java_frames_include_no_semicolons(
...
+        monkeypatch.setattr(AsyncProfiledProcess, "stop_async_profiler", flush_output_and_stop_async_profiler)
+
+    with make_java_profiler(
+        storage_dir=str(tmp_path),
+        duration=3,
+        frequency=999,
+    ) as profiler:
+        collapsed = snapshot_pid_collapsed(profiler, application_pid)
+        assert_collapsed(collapsed)
+        application_docker_container.stop(timeout=3)
```
[the remark this answers is not in the document]
Comment (marcin-ol) [reply tests/test_java.py:756]: Thanks, will fix it.

File: tests/test_java.py (hunk 2 above)
Comment (marcin-ol) [reply tests/test_java.py]: 👍

**Comment by marcin-ol** [comment]
> Done reviewing.
> 
> Please run a negative test - i.e a test where you change the AP reference to your fork, which removes e.g [Granulate/async-profiler@35531a4](https://github.com/Granulate/async-profiler/commit/35531a45a197048a450aca38386d144dfa1060bb), so it proves that the test now fails.

Added a test commit to async-profiler bringing back the prints that were removed: [marcin-ol/async-profiler@623637b](https://github.com/marcin-ol/async-profiler/commit/623637b593282cc7fe5a07b85d35c00bc6c245d1).
- engaging the test against that commit: [gprofiler@9343ff4](https://github.com/Granulate/gprofiler/pull/664/commits/9343ff484ca0ce3d43c7fc6de83211004e5c1246)
 
I expect the test not to fail - although it should - when flushing is disabled.  
Introduced this commit to back up this claim.

**Comment by marcin-ol** [comment]
> > Done reviewing.
> > Please run a negative test - i.e a test where you change the AP reference to your fork, which removes e.g [Granulate/async-profiler@35531a4](https://github.com/Granulate/async-profiler/commit/35531a45a197048a450aca38386d144dfa1060bb), so it proves that the test now fails.
> 
> Added a test commit to async-profiler bringing back the prints that were removed: [marcin-ol/async-profiler@623637b](https://github.com/marcin-ol/async-profiler/commit/623637b593282cc7fe5a07b85d35c00bc6c245d1).
> 
> * engaging the test against that commit: [gprofiler@9343ff4](https://github.com/Granulate/gprofiler/pull/664/commits/9343ff484ca0ce3d43c7fc6de83211004e5c1246)
> 
> I expect the test not to fail - although it should - when flushing is disabled. Introduced this commit to back up this claim.

Without flushing async-profiler's output, the test fails, even though the profiler emitted additional output:
```
tests/test_java.py::test_no_stray_output_in_stdout_stderr[False-True] PASSED [ 35%]
tests/test_java.py::test_no_stray_output_in_stdout_stderr[True-True] FAILED [ 36%]
```

**Comment by Jongy** [comment]
> Without flushing async-profiler's output, the test fails, even though the profiler emitted additional output:
> 
> ```
> tests/test_java.py::test_no_stray_output_in_stdout_stderr[False-True] PASSED [ 35%]
> tests/test_java.py::test_no_stray_output_in_stdout_stderr[True-True] FAILED [ 36%]
> ```

If I understand it correctly - it looks good - the test fails when we DO flush and passes when we don't.

That's what I meant by a negative test. We can remove the async-profiler build commit change, it looks ready for merge IMO.

**Comment by marcin-ol** [comment]
> > Without flushing async-profiler's output, the test fails, even though the profiler emitted additional output:
> > ```
> > tests/test_java.py::test_no_stray_output_in_stdout_stderr[False-True] PASSED [ 35%]
> > tests/test_java.py::test_no_stray_output_in_stdout_stderr[True-True] FAILED [ 36%]
> > ```
> 
> If I understand it correctly - it looks good - the test fails when we DO flush and passes when we don't.
> 
> That's what I meant by a negative test. We can remove the async-profiler build commit change, it looks ready for merge IMO.

Yeah, that's what should be expected.
I've made flushing nonvoluntary and restored reference to mainstream async-profiler.

**Review by Jongy — APPROVED** [review]

**Comment by Jongy** [comment]
Awesome @marcin-ol :)
